### ISSUE TITLE
No autocompletion in .md files in Textastic - Does the Shortcut still work?

### ISSUE THREAD

**Issue by FeralFlora** [somelinguist/Textastic-Citation-Completions#1]
I tried installing the Shortcut following the first guide, but even though I get through the setup without issues, there are no autocompletions options triggered when I type `@` in Textastic? Do the script and the shortcut still work?

**Comment by somelinguist** [comment]
Hi, I haven't used the shortcut myself in a few years, but I was able to get it to work again just now.

Things to check:
1. During setup, when it asks to create the `citations.js` file you have to navigate in the file picker to `LOCALDEVICE/Textastic/#Textastic/CodeCompletion/`, where `LOCALDEVICE` is either `On My iPhone` or `On My iPad`. You may have to create all of those subdirectories n Textastic. The shortcut itself doesn't care where you choose...it will just output the file to wherever you choose. However, Textastic will only pick up the completions if it's saved at that location (see https://www.textasticapp.com/v9/manual/customization/custom_syntax_themes_templates.html and https://github.com/blach/Textastic-Customization#code-completion).

2. If you don't choose a default .bib file during setup, you'll need to later set a .bib file by using the share sheet to send it to the shortcut.

3. Make sure the markdown file you're editing is set to Markdown syntax in Textastic (it should be by default, but I've changed it a few times myself inadvertently).

If you've done the above, check the `LOCALDEVICE/Textastic/#Textastic/CodeCompletion/citations.json` file. It should start off something like this:


```json
{
  "description": "Code Completion for Citations",
  "uuid": "6a6b74bb-499d-43cf-89f7-8a2d0815d0d8",
  "completionSets": [
    {
      "name": "citation.references",
      "defaultAppend": "",
      "strings": [
        {
          "string": "lambrecht1997 | Typological Variation in Sentence-Focus Constructions",
          "replace": "lambrecht1997"
        },
        {
          "string": "lambrecht1994 | Information Structure and Sentence Form: Topic, Focus, and the Mental Representations of Discourse Referents",
          "replace": "lambrecht1994"
        },
```

If the file isn't there or doesn't look similar:

1. In Scriptable's settings, go to File Bookmarks and delete any bookmarks that start with `Textastic Citation`
2. Delete the file `/iCloud Drive/Shortcuts/citationSetup.json`
3. Then you can run the shortcut again to go through setup again.


If the file is there and looks similar, restarting Textastic or going to Textastic's settings and tapping `Other->Reload Customizations` might make it work. This shouldn't be necessary in the last version of the shortcut, as Textastic's developer had added a callback url that the shortcut uses to do this automatically.

Let me know if that helps!

**Comment by FeralFlora** [comment]
Thanks a lot for that comprehensive reply! I will try your suggested solutions. I think the issue might be the directory, I don't think it did it like that, so I will try to redo it.

**Comment by FeralFlora** [comment]
I got it working now! You were right, the problem was the path for the citations.json file. I had merely put it in the base directory of Textastic.

By the way, since you mentioned that you haven't used this shortcut for a few years, I'm wondering if that's because you've found a better solution for citations / academic writing on an iPad? Such workflows are very hard too come by, so I'm glad I found your shortcut and also eager to hear about anything else out there.

**Comment by FeralFlora** [comment]
Another quick question, do you think it would be possible to use a modifier key or something like that to insert an alternative citekey format, like `[@citekey]`?

**Comment by somelinguist** [comment]
I'm glad you were able to get it to work.

Regarding not using it any more: Unfortunately, though I'd like to, I haven't don't much academic writing on my iPad because there isn't a good way to insert special characters or non-standard diacritics using an external keyboard. While Apple makes it possible for developers to create custom on screen keyboards, there isn't a way, as of yet, to create custom layouts for external keyboards like there is on Mac or Windows. 

And as a linguist, a lot of my writing requires being able to insert such characters diacritics repeatedly. So I mostly end up writing everything on my Mac. I'd love to be able to use my iPad for it though. 

I do take a lot of notes on my phone or iPad when researching. More recently I've been using Drafts for such notes, mostly because I don't have to think about what to name a file or where to put them, and I can organize later. If I were able to do more actual long-form writing, I'd love to still use Textastic or Taio for markdown.

If you use LaTeX, Texifier (formerly TeXPad) does a pretty good job for inserting citations.

**Comment by somelinguist** [comment]
> Another quick question, do you think it would be possible to use a modifier key or something like that to insert an alternative citekey format, like `[@citekey]`?

I thought about this, but if I remember correctly, Textastic just uses the inputted text for suggesting completions, and doesn't take into account actually keystrokes.

For `[@citekey]`, I've just entered the brackets myself, as `@` still triggers the completion. If I remember right, that was also the best way to support entering multiple citations in the same unit.

If you want, you can change the script that Scriptable uses to generate the completions file here:

https://github.com/somelinguist/Textastic-Citation-Completions/blob/e75cf2b5b78cf9994dd573628357dcba188b18f4/shortcut/Update%20Textastic%20Citations.js#L165-L170

At the end of line 168, I think you could change `"replace": match[2]` to something like `"replace": "[" + match[2] + "]"`. However, that would change all of the inserts.

The shortcut/script also offers citation completions for LaTeX files using `\cite{`, etc for the triggers. So changing the code as above would probably break that too.

**Comment by FeralFlora** [comment]
> I'm glad you were able to get it to work.
> 
> Regarding not using it any more: Unfortunately, though I'd like to, I haven't don't much academic writing on my iPad because there isn't a good way to insert special characters or non-standard diacritics using an external keyboard. While Apple makes it possible for developers to create custom on screen keyboards, there isn't a way, as of yet, to create custom layouts for external keyboards like there is on Mac or Windows.
> 
> And as a linguist, a lot of my writing requires being able to insert such characters diacritics repeatedly. So I mostly end up writing everything on my Mac. I'd love to be able to use my iPad for it though.
> 
> I do take a lot of notes on my phone or iPad when researching. More recently I've been using Drafts for such notes, mostly because I don't have to think about what to name a file or where to put them, and I can organize later. If I were able to do more actual long-form writing, I'd love to still use Textastic or Taio for markdown.
> 
> If you use LaTeX, Texifier (formerly TeXPad) does a pretty good job for inserting citations.

Okay, thanks for the information. I see, yeah that sounds quite limiting for your field. Myself, I am working in Obsidian, but as it doesn't support citation on mobile devices yet, I was looking for ways to edit my files with something else capable of inserting pandoc markdown citations. 

I also found an old Shortcut for inserting citekeys from Zotero using the web API, which partially works, but it requires you to pin all your citekeys for them to be available online (this adds them to the Extra field). See here: https://www.icloud.com/shortcuts/9567bf9b62464a2ba253043598757665 

This solution works across all apps. My citekeys weren't pinned when testing it, so I couldn't verify that it works, but it does access the API and return search results as expected. Another issue is that the search repeats every title for every author. So, an article with five authors will be present five times in the list of results.

I have used latex before, but for my current work, I am aiming to stick with markdown, but thanks for the recommendation.

**Comment by FeralFlora** [comment]
This archive link contains an explainer about the Shortcut above. You have to scroll to find it. I don't know if any of this is relevant for you, because you might not use Zotero. https://web.archive.org/web/20230201050309/https://appademic.tech/

**Comment by FeralFlora** [comment]
> > Another quick question, do you think it would be possible to use a modifier key or something like that to insert an alternative citekey format, like `[@citekey]`?
> 
> I thought about this, but if I remember correctly, Textastic just uses the inputted text for suggesting completions, and doesn't take into account actually keystrokes.
> 
> For `[@citekey]`, I've just entered the brackets myself, as `@` still triggers the completion. If I remember right, that was also the best way to support entering multiple citations in the same unit.
> 
> If you want, you can change the script that Scriptable uses to generate the completions file here:
> 
> https://github.com/somelinguist/Textastic-Citation-Completions/blob/e75cf2b5b78cf9994dd573628357dcba188b18f4/shortcut/Update%20Textastic%20Citations.js#L165-L170
> 
> At the end of line 168, I think you could change `"replace": match[2]` to something like `"replace": "[" + match[2] + "]"`. However, that would change all of the inserts.
> 
> The shortcut/script also offers citation completions for LaTeX files using `\cite{`, etc for the triggers. So changing the code as above would probably break that too.

Yeah, it's not too much of a hassel to type the brackets myself. But as I normally use that format, it might be best to change the script so it is the default.

Apropos changing the script, I have one last question, I hope that's okay. The thing is that I am working in a markdown variant calle Quarto (.qmd), which is a successor to rmarkdown made by Rstudio. Like I wrote, quarto files have a .qmd file extension. I can edit these files in Textastic just fine, but the autocompletion doesn't work, because the scope only targets markdown files. Do you know if it would be possible to extend the scope to include Quarto files with the .qmd extension? Or is that limited by what Textastic supports?

**Comment by somelinguist** [comment]
> I also found an old Shortcut for inserting citekeys from Zotero using the web API, which partially works, but it requires you to pin all your citekeys for them to be available online (this adds them to the Extra field). See here: https://www.icloud.com/shortcuts/9567bf9b62464a2ba253043598757665
> 
> This solution works across all apps. My citekeys weren't pinned when testing it, so I couldn't verify that it works, but it does access the API and return search results as expected. Another issue is that the search repeats every title for every author. So, an article with five authors will be present five times in the list of results.


That sounds like an interesting shortcut. I think I might have used something similar before. I had been using Zotero until recently as well, with the Better BibTex extension, which would pin my (on desktop at least?). I had the extension set up to automatically export a bibtex file of my library to an iCloud folder, and also for Zotero collections that I used for projects. Then I could use the Textastic shortcut here to set completions to the .bib file I needed to use.

Recently I switched to using Bookends instead of Zotero, however. This was mostly because I was using linked attachments synced via iCloud instead of syncing them via Zotero. I wanted to be able to use a third-party PDF app to read and annotate attachments.  I could never get a workflow that I liked with Zotero for adding new references/PDFs on iPad or iPhone, or for using Zotero to find and then open those PDFs on my iPad in my PDF app. I do really like Zotero, I just couldn't get it to work exactly how I wanted. Bookends, while not free, has really improved that particular workflow for me.

**Comment by somelinguist** [comment]
> Apropos changing the script, I have one last question, I hope that's okay. The thing is that I am working in a markdown variant calle Quarto (.qmd), which is a successor to rmarkdown made by Rstudio. Like I wrote, quarto files have a .qmd file extension. I can edit these files in Textastic just fine, but the autocompletion doesn't work, because the scope only targets markdown files. Do you know if it would be possible to extend the scope to include Quarto files with the .qmd extension? Or is that limited by what Textastic supports?

Are you using a custom sytnax file for Textastic to get syntax highlighting for the .qmd files? If, not do you have to choose Markdown as the syntax for your files manually in Textastic?

If you're using a custom syntax, you can add whatever the scope the syntax uses to the script at line 32, and I think it should work?

https://github.com/somelinguist/Textastic-Citation-Completions/blob/e75cf2b5b78cf9994dd573628357dcba188b18f4/shortcut/Update%20Textastic%20Citations.js#L32

If you don't have to change the syntax manually in Textastic, I'm not sure what the problem might be.

You can get Textastic to show you what scopes apply to the cursor position with this keyboard shortcut: Ctrl + Shift + P. You can add whatever the most basic scope for the file is to the line above.

**Comment by somelinguist** [comment]
You might also be interested in the other completions I did for Pandoc here: https://github.com/somelinguist/Textastic-Pandoc-Markdown-Completions

To get I to work with the .qmd files, you might need to add whatever scope the files are using to the relevant `scope` values in lines 160-215.

https://github.com/somelinguist/Textastic-Pandoc-Markdown-Completions/blob/063f20665651d316913ede7f37fa923fd62b9496/pandoc.json#L160-L215

**Comment by FeralFlora** [comment]
> > Apropos changing the script, I have one last question, I hope that's okay. The thing is that I am working in a markdown variant calle Quarto (.qmd), which is a successor to rmarkdown made by Rstudio. Like I wrote, quarto files have a .qmd file extension. I can edit these files in Textastic just fine, but the autocompletion doesn't work, because the scope only targets markdown files. Do you know if it would be possible to extend the scope to include Quarto files with the .qmd extension? Or is that limited by what Textastic supports?
> 
> Are you using a custom sytnax file for Textastic to get syntax highlighting for the .qmd files? If, not do you have to choose Markdown as the syntax for your files manually in Textastic?
> 
> If you're using a custom syntax, you can add whatever the scope the syntax uses to the script at line 32, and I think it should work?
> 
> https://github.com/somelinguist/Textastic-Citation-Completions/blob/e75cf2b5b78cf9994dd573628357dcba188b18f4/shortcut/Update%20Textastic%20Citations.js#L32
> 
> If you don't have to change the syntax manually in Textastic, I'm not sure what the problem might be.
> 
> You can get Textastic to show you what scopes apply to the cursor position with this keyboard shortcut: Ctrl + Shift + P. You can add whatever the most basic scope for the file is to the line above.

Actually, Textastic was just considering the Quarto file as plain text, so I had no syntax highlighting, but could edit. I turns out, all I needed to do to make the citation autocompletion work in the .qmd file was choose markdown as the syntax of the file. After that, it works!

I tried editing the line you mentioned before my previous message, but it was just on a hunch `text.qmd.quarto`, but that didn't work. But thanks for the tip about how to see the scope at a line.

**Comment by FeralFlora** [comment]
> You might also be interested in the other completions I did for Pandoc here: https://github.com/somelinguist/Textastic-Pandoc-Markdown-Completions
> 
> To get I to work with the .qmd files, you might need to add whatever scope the files are using to the relevant `scope` values in lines 160-215.
> 
> https://github.com/somelinguist/Textastic-Pandoc-Markdown-Completions/blob/063f20665651d316913ede7f37fa923fd62b9496/pandoc.json#L160-L215

Yes, I saw that, and I'm going to try it out. Changing the syntax in the Textastic menu changes the scope (found with ctrl + shift + p), so I shouldn't be necessary to add anything.

**Comment by FeralFlora** [comment]
One limitation with the autocompletion that I've found is that the search matches to the beginning of the file, so you can only search for first authors, not titles. I wish it was a fuzzy search, but I don't know if it's possible to change that.

**Comment by somelinguist** [comment]
> One limitation with the autocompletion that I've found is that the search matches to the beginning of the file, so you can only search for first authors, not titles. I wish it was a fuzzy search, but I don't know if it's possible to change that.

Yes, I think that would be helpful too. The matching just uses regular expressions. The markdown one is here:

https://github.com/somelinguist/Textastic-Citation-Completions/blob/e75cf2b5b78cf9994dd573628357dcba188b18f4/shortcut/Update%20Textastic%20Citations.js#L33

Right now, roughly, it just looks for any text following `@` that matches one of the completion records. It's set to use the completion index at (1), which is `.*` in the file.

I'm not sure what could be changed to make the match in a more fuzzy manner.
